tests/subsys/usb_c/sink/src/main.c

```diff
@@ -0,0 +1,514 @@
+	data->pn_hard_reset_received = ATOMIC_INIT(false);
+	data->pn_source_capabilities_received = ATOMIC_INIT(false);
+	data->pn_sender_response_timeout = ATOMIC_INIT(false);
+
+	data->uut_received_src_cap_num = 0;
+	data->snk_cap_pdo.raw_value = 0;
+	data->rx_msg.len = 0;
+
+	/* Initialize PDO for sending in step 5 */
+
+	/* a) B31…30 (Fixed Supply) set to 00b */
+	pdo.type = PDO_FIXED;
+	/* b) B29 (Dual-Role Power) set to 1b */
+	pdo.dual_role_power = 1;
+	/* c) B28 (USB Suspend Supported) set to 0b */
+	pdo.usb_suspend_supported = 0;
+	/* d) B27 (Unconstrained Power) set to 1b */
+	pdo.unconstrained_power = 1;
+	/* e) B26 (USB Communications Capable) set to 0b */
+	pdo.usb_comms_capable = 0;
+	/* f) B25 (Dual-Role Data) set to 0b */
+	pdo.dual_role_data = 0;
+	/* g) B24 (PD3, Unchunked Extended Messages Supported) set to 0b */
+	pdo.unchunked_ext_msg_supported = 0;
+	/*
+	 * h) B23 (EPR Mode Capable) to 0b, unless it is mentioned in the
+	 *    test procedure. NOTE: NOT CURRENTLY SUPPORTED IN THE SUBSYSTEM.
+	 */
+	pdo.reserved0 = 0;
+	/* i) B21…20 (Peak Current) set to 00b */
+	pdo.peak_current = 0;
+	/* j) B19…10 (Voltage) set to 5V */
+	pdo.voltage = PD_CONVERT_MV_TO_FIXED_PDO_VOLTAGE(5000);
+	/* k) B9…0 (Maximum Current) set to 100mA */
+	pdo.max_current = PD_CONVERT_MA_TO_FIXED_PDO_CURRENT(100);
+
+	/* 1) The test starts in a disconnected state. */
+	tester_disconnected();
+	/* Give the Sink state machine time to transition */
+	k_msleep(20);
+
+	/* 2) Apply Rp */
+	tester_apply_cc(TC_CC_VOLT_RP_3A0, TC_CC_VOLT_OPEN);
+
+	/* 3) Apply vSafe5V on VBUS. */
+	tester_apply_vbus(PD_V_SAFE_5V_MIN_MV);
+
+	/* 4) The Tester waits until TC_ATTACHED_SNK state is reached. */
+	for (i = 0; i < 500; i++) {
+		/* NOT_PD_CONNECTED notification is sent when the PE starts up */
+		if (atomic_get(&data->pn_not_pd_connected)) {
+			break;
+		}
+		k_usleep(500);
+	}
+
+	/* 5) Send up to 50 source capability messages to UUT */
+	for (i = 0, j = 0; i < 50; i++) {
+		switch (j) {
+		case 0:
+			/* Send Source Cap message */
+			tester_send_data_msg(PD_DATA_SOURCE_CAP, &pdo.raw_value, 1, true);
+			j++;
+			k = 0;
+			break;
+		case 1:
+			/* Wait up to 50 mS for the UUT to receive the Source Cap message */
+			if (atomic_get(&data->pn_source_capabilities_received)) {
+				/* The UUT received the Source Cap Message, so exit loop */
+				i = 50;
+			} else if (++k == 50) {
+				/* The UUT didn't detect the Source Cap message. Send it again */
+				j = 0;
+			}
+			k_usleep(500);
+			break;
+		}
+	}
+	zassert_true(atomic_get(&data->pn_source_capabilities_received),
+		     "UUT didn't receive Source Caps message");
+
+	/* 6) Wait until UUT processes the Source Cap message */
+	for (i = 0; i < 500; i++) {
+		if (atomic_get(&data->pn_pd_connected)) {
+			break;
+		}
+		k_usleep(500);
+	}
+
+	/* 7) The check fails if the UUT does not respond with a Request message. */
+	zassert_true(atomic_get(&data->uut_request), "UUT didn't send request message.");
```

Comment:
This seems slightly incorrect:
1. `uut_request` gets set from the notify call from the stack to the DPM, whereas the result of `tester_get_uut_tx_data` is set by the tester when it actually receives the message. So you're taking the stack's word for it that it sent the message instead of having the tester verify that it received it.
2. `uut_request` gets set when the stack gets the RDO from the DPM, which it needs to do before it sends the Request message. So there will be at least some amount of time when `uut_request` is true, but the message hasn't been sent yet. In principle, the stack could request the RDO and never send the message, and the received message data would be invalid.

It seems like a more appropriate flow would be to poll the tester until it has received a message (step 6 above) and then check that that message is a Request and then check the RDO fields as below.